### DISCUSSION TITLE
Database encryption

### DISCUSSION THREAD

**Discussion by fbede** [pocketbase/pocketbase#1961] [Q&A] [answered]
Hey, I'm just curious so I have a series of questions. Before I ask I know that this is the best MVP alternative to quickly get something up and running and nobody is or should be using pocketbase for mission critical tasks especially since it is still in beta. So here are my questions...

1. Is there a way to encrypt the data stored in the database on pocketbase?
2. Is everything visible if the server gets hacked?
3. Is it even safe to run a production server without the database content encrypted?
4. Is requiring end to end encryption on a personal finance tracker app asking for two much?

I thought of these questions because I'm building a personal finance tracker and I was thinking of encrypting the embedded database since it's an offline first app. The problem is if the backend is not also encrypted the security problem of keeping users data safe is not exactly solved. Please I would like to hear opinions on this. Thank you.

**Comment by ganigeorgiev** [comment, marked as answer]
> 1. Is there a way to encrypt the data stored in the database on pocketbase?

There is no built-in encryption for the db content, if that's what you are asking. You can encrypt the entire db file manually using external tools like `gpg` but it is up to the developer to manage the secret key and decrypt it on server start. Alternatively, when [using as framework](https://pocketbase.io/docs/use-as-framework/), you can encrypt individual field values, but again it is up to the developer to handle it on their own if they really need it.

---

> 2. Is everything visible if the server gets hacked?

I'm not sure that I understand your question. If your server get hacked, everything what is on the server is compromised. This is not unique to PocketBase.

Access restrictions to the PocketBase API and the underlying data is done via the collection API rules (by default only admins have access). You can read more in https://pocketbase.io/docs/api-rules-and-filters/.

---

> 3. Is it even safe to run a production server without the database content encrypted?

Once again, I'm not sure that I understand the question and what do you mean by _"without the database content encrypted"_. Encryption at what stage - at rest, column/field level, full-disk encryption? Most of the web applications don't store their db content encrypted.

---

> 4. Is requiring end to end encryption on a personal finance tracker app asking for two much?

If you are looking for end-to-end encrypted db solution, then PocketBase may not be the right choice for your app. You have to decide what features set you want to support because end-to-end encryption doesn't come without cost:
- db encryption at what level? do you really need encryption for the entire database or only some specific table/column?
- will there be a central encryption key or only the users will have access to their data (eg. via their password as key), meaning that even admins will not be able to read/modify it?
- if an individual key is used, how the relations between the encrypted users data will be managed?
- how the data querying will be performed? (_it is doable for simple equal type of comparisons but for anything more complex you'll have to decrypt the data at some point and you'll hit performance bottlenecks_)
- how big the encrypted data could get and where will be loaded for read/modify - in the runtime memory?, etc.

**Reply by fbede** [reply]
Thank you so much. Your reply was faster than I expected. Like I said earlier, I was just curious and thinking aloud... Encrypting the whole database was what I was asking about but your reply already helped. Thank you.

**Reply by oSethoum** [reply]
@fbede  Hi, if you wanna encrypt the database files using sqlcipher there is a package for that, what I did is cloning the `pocketbase` repo and edited the part where the connection to the database is made it will be in file `core/db_cgo.go` 
in this case you can change the secret manually, that's the only downside of it in my opinion. 
```go
//go:build cgo

package core

import (
	"github.com/pocketbase/dbx"
	// _ "github.com/mattn/go-sqlite3"
	_ "github.com/mutecomm/go-sqlcipher/v4"
)

func connectDB(dbPath string) (*dbx.DB, error) {
	db, err := dbx.Open("sqlite3", dbPath+"?_pragma_key=secret&_pragma_cipher_page_size=4096")
	if err != nil {
		return nil, err
	}

	if err := initPragmas(db); err != nil {
		db.Close()
		return nil, err
	}

	return db, nil
}

```

**Reply by ganigeorgiev** [reply]
@oSethoum _Note that this works only for the CGO version (aka. if you are building with `CGO_ENABLED=1`)._

**Reply by oSethoum** [reply]
@ganigeorgiev Thank you, I forgot to add that.

**Reply by zclcdco** [reply]
Hi there, is this still the recommended method for using SQLCipher on Pocketbase? Are there any other easier ways than cloning the entire repo? It seems like it might make it less maintainable if this process has to be repeated every time there is an update...

**Reply by ganigeorgiev** [reply]
@zclcdco With the [next major release](https://github.com/pocketbase/pocketbase/discussions/4355) it will be possible to pass an external `connectDB` function to the app instance (there are no ETAs yet). Another alternative without forking could be to use the `//go:linkname` directive and overwrite the `github.com/pocketbase/pocketbase/core.connectDB` symbol with your own function.

**Comment by erolisdev** [comment]
It's great that Pocketbase allows adding a custom SQL driver. [Example](https://pocketbase.io/docs/go-overview/#github-commattngo-sqlite3)
Can you recommend an up-to-date SQLCipher go package that I can use with Pocketbase?
